flox/xarray.py

```diff
@@ -413,19 +417,19 @@ def wrapper(array, *by, func, skipna, **kwargs):
     if unindexed_dims:
         actual = actual.drop_vars(unindexed_dims)
 
-    if len(by) == 1:
+    if len(by_broad) == 1:
         for var in actual:
-            if isinstance(obj, xr.DataArray):
-                template = obj
-            else:
+            if isinstance(obj, xr.Dataset):
                 template = obj[var]
+            else:
+                template = obj
             if actual[var].ndim > 1:
-                actual[var] = _restore_dim_order(actual[var], template, by[0])
+                actual[var] = _restore_dim_order(actual[var], template, by_broad[0])
 
     if missing_dim:
         for k, v in missing_dim.items():
             missing_group_dims = {
-                dim: size for dim, size in group_sizes.items() if dim not in v.dims
```

Comment:
lol sorry for this mess :) I like redefining stuff.

Reply:
Many times it looks cleaner redefining, but mypy is more sceptical. :) Especially when the type is changed or narrowed. I'll ponder on better names a little bit.